### PR TITLE
Reworking CPU data queries and access for vmvx/builtins/compiler-produced code.

### PR DESCRIPTION
**Pull request by benvanik**
This partially reverts #10249 and instead passes the processor ID through to a worker-specific VMVX module state. The VM side shouldn't really need the processor ID while the VMVX side does and this way we don't need to perform complex pass-throughs.

In addition CPU data (features, values, etc) is now centralized and specified in a header that can be shared across all potential consumers. The storage of the CPU data cannot cross the builtins barrier but can be queried directly from the VMVX module and then passed into the builtins library as an argument. Now any VMVX function defined in the vmvx module.c can now use `iree_cpu_data_fields()` to get the bits and the `IREE_CPU_DATA_FIELD_*` macros to test the bits, or pass the fields into the ukernel builtins and let that test the bits.

Since we aren't stable yet we're free to change the fields for now - what's here can be considered a template.

### CODE REVIEW

File: runtime/src/iree/modules/vmvx/module.c
Comment (ScottTodd) [inline runtime/src/iree/modules/vmvx/module.c:50]: > Since we aren't stable yet we're free to change the fields for now - what's here can be considered a template.

FYI this buildkite pipeline tests the latest release and latest source code together and it spotted a break earlier from the original PR: https://buildkite.com/iree/iree-samples/builds/460#0182f142-c462-4fa8-8963-b653bf4ef7a6
```
[2022-08-31T00:36:06Z] + /usr/src/github/iree/build-samples/samples/variables_and_state/variables-and-state /tmp/iree/colab_artifacts/counter_vmvx.vmfb local-task
[2022-08-31T00:36:06Z] ./samples/variables_and_state/test.sh: line 31: 14169 Segmentation fault      (core dumped) ${BUILD_DIR}/samples/variables_and_state/variables-and-state ${ARTIFACTS_DIR}/counter_vmvx.vmfb local-task
```
(that resolved itself after the next release brought the two back in sync)

**Review by ScottTodd — APPROVED** [review]

File: runtime/src/iree/base/internal/cpu.c
Comment (bjacob) [inline runtime/src/iree/base/internal/cpu.c:37]: discrepancy between just `ASIMD` (another name for NEON in general) and `ASIMDDP` (which means dot-product extension)

File: runtime/src/iree/base/internal/cpu.c
Comment (bjacob) [inline runtime/src/iree/base/internal/cpu.c:39]: If we rely on explicit `HWCAP` bits to detect such recent CPU features then we risk running again into the trouble we were in with the Pixel4 launch where it was only weeks ahead of the public device launch that an OS image update brought a Linux kernel backport exposing the `HWCAP` feature bit for the `DP` extension.

My lesson learned from that episode is that it's better to rely only on `HWCAP_CPUID` and use trapped access to CPU id registers (https://docs.kernel.org/arm64/cpu-feature-registers.html).  This still relies a sufficiently recent kernel version to have `HWCAP_CPUID`, but at least from there on, the kernel version requirement is frozen, so for cutting-edge features such as `i8mm` that is a win. As the Pixel4 experience showed, Android kernel versions on brand new Pixel devices can happily trail upstream by years.

https://github.com/iree-org/iree/blob/main/runtime/src/iree/tooling/cpu_features.c shows an example of using the `HWCAP_CPUID` way.

It's totally cool if you're OK with the idea but still want to merge this as-is and defer to me to do that as a follow-up - just confirm if we're aligned on the strategy :-)

Comment (bjacob) [reply runtime/src/iree/base/internal/cpu.c:39]: I can also live with the explicit `HWCAP` feature flags for now as long as my moto device has a recent enough kernel (i need to check) and as long as we revisit before someone tries to ship something to real users (some of whom will be running a kernel not exposing useful features in `HWCAP` bits)

File: runtime/src/iree/base/internal/cpu.c
Comment (bjacob) [inline runtime/src/iree/base/internal/cpu.c:82]: There are multiple sets of strings in existence to refer to these features. Can we stick to LLVM (equivalently, GCC) cpu-features strings that are what we need to pass to the compiler? The string for dot-product instructions in "dotprod".   For i8mm, it stays "i8mm".

File: runtime/src/iree/base/internal/cpu.c
Comment (bjacob) [inline runtime/src/iree/base/internal/cpu.c:79]: While the ability to look up by string key is potentially useful, for my immediate need in builtins I'm better off looking up by symbolic constant.

File: runtime/src/iree/base/internal/cpu.h
Comment (bjacob) [inline runtime/src/iree/base/internal/cpu.h:35]: I don't understand what this ability to export raw cached data is needed for. If we only had opaque `iree_cpu_initialize` and symbolic-constant-based bit lookup, wouldn't that be enough and avoid the need to allocate a somewhat larger, arbitrary-sized `iree_cpu_data_cache_` ?

File: runtime/src/iree/hal/local/executable_environment.c
Comment (bjacob) [inline runtime/src/iree/hal/local/executable_environment.c:30]: do we really need to have the cpu data stored twice in memory - in `out_environment->processor.data` and in `iree_cpu_data_cache_` ?

File: runtime/src/iree/hal/local/local_executable_cache.c
Comment (bjacob) [inline runtime/src/iree/hal/local/local_executable_cache.c:18]: It's not clear what the relationship is between the parts of this diff that refer to `worker_capacity` and the rest. Should that be a separate PR?

File: runtime/src/iree/vm/stack.c
Comment (bjacob) [inline runtime/src/iree/vm/stack.c:233]: What is the relationship between `iree_vm_stack_reset` parts and other parts of this PR?

File: runtime/src/iree/base/internal/cpu.c
Comment (bjacob) [reply runtime/src/iree/base/internal/cpu.c:79]: I can see that it's useful in `iree_hal_inline_module_device_query_i64`, but that's not my primary use case, since I want to query cpu info from C code.

**Review by bjacob — CHANGES_REQUESTED** [review]
On closer look, there are enough open discussion topics here that I'm selecting "request changes" to make sure we at least chat a bit about them :-)

File: runtime/src/iree/base/internal/cpu.h
Comment (benvanik) [reply runtime/src/iree/base/internal/cpu.h:35]: You cannot call these functions from the ukernel library or generated code and the fields need to be passed in to such code in order to be used. This allows access to the fields so that they can be passed in.

File: runtime/src/iree/hal/local/local_executable_cache.c
Comment (benvanik) [reply runtime/src/iree/hal/local/local_executable_cache.c:18]: Check the commit descriptions when reviewing PRs with multiple commits :)
In order to pass in per-worker state like the processor ID (as you asked for) per-worker state needed to be allocated and that requires knowing how many workers there are.

File: runtime/src/iree/vm/stack.c
Comment (benvanik) [reply runtime/src/iree/vm/stack.c:233]: Cleanup as part of what I was doing.

File: runtime/src/iree/base/internal/cpu.c
Comment (benvanik) [reply runtime/src/iree/base/internal/cpu.c:82]: Sure - if you have a table that can be referenced and how they map that'd be useful.

Comment (bjacob) [reply runtime/src/iree/base/internal/cpu.c:82]: For aarch64 it's in this file: https://github.com/llvm/llvm-project/blob/main/llvm/lib/Target/AArch64/AArch64.td#L295

File: runtime/src/iree/base/internal/cpu.c
Comment (bjacob) [reply runtime/src/iree/base/internal/cpu.c:39]: as noted above I'm OK to refactor/merge/dedupe that as a followup after the present PR is merged, if we agree on the approach.

File: runtime/src/iree/base/internal/cpu.c
Comment (bjacob) [reply runtime/src/iree/base/internal/cpu.c:82]: That link doesn't give the mapping with the other set of strings, though. The best I'm able to find is https://github.com/llvm/llvm-project/blob/main/llvm/lib/Target/AArch64/AArch64InstrInfo.td giving the mapping between LLVM feature strings and instruction mnemonics.

The other set of strings like `asimddp` it's not even 100% clear who owns it. These strings are somewhat but not perfectly in agreement with the ARM reference manual. For example, in the ARM reference manual it's `DP` not `asimddp`.  I actually believe that the strings like `asimddp` are Linux specific.  The complete mapping between the two sets of strings might be as complex as this 4-deep chain:

```
LLVM strings like "dotprod"
^
| https://github.com/llvm/llvm-project/blob/main/llvm/lib/Target/AArch64/AArch64InstrInfo.td
V
Instruction mnemonics like "sdot"
^
| ARMv8-a architecture reference manual (online PDF from ARM.com)
| Description of each instruction like "sdot"
V
ARM feature identification mnemonics like "DP"
^
| ARMv8-a architecture reference manual (online PDF from ARM.com)
| Description of control registers bitfield entries like "DP"
V
ARM feature identification bit values
^
| Linux kernel implementation of /proc/cpuinfo and `HWCAP` - usually in same patch set.
V
Linux strings like "asimddp"
```

File: runtime/src/iree/base/internal/cpu.c
Comment (benvanik) [reply runtime/src/iree/base/internal/cpu.c:45]: I'm going to leave it for the moment because I don't have a system to test on, but I did add a TODO noting that we should make use of it!

Comment (bjacob) [reply runtime/src/iree/base/internal/cpu.c:45]: SG - i'll take care of it next week.

File: runtime/src/iree/base/internal/cpu.c
Comment (benvanik) [reply runtime/src/iree/base/internal/cpu.c]: I really like your suggestion of following LLVM here - will make it really easy to track across cmake, the compiler/codegen, user command lines for iree-compile, and the runtime!

**Review by bjacob — APPROVED** [review]

**Pull request merged**
